File: provider/Dignitas.Specmatic.Provider.API/Services/CartService.cs
using Dignitas.Specmatic.Provider.API.Exceptions;
using Dignitas.Specmatic.Provider.API.Models;

namespace Dignitas.Specmatic.Provider.API.Services;

public class CartService
{
    private static readonly Dictionary<uint, Cart> Carts = new()
    {
        {
            0,
            new Cart
            {
                CartId = 0,
                UserId = 409,
                Items = [ new CartItem {
                    ProductId = 9,
                    Quantity = 1
                }]
            }
        }
    };
    private static uint _cartIdCounter = 1;

    private readonly UserService _userService;
    private readonly ProductService _productService;

    public CartService(UserService userService, ProductService productService)
    {
        _userService = userService;
        _productService = productService;
    }

    public InitializeCartResponse InitializeCart(InitializeCartRequest request)
    {
        if (!_userService.Exists(request.UserId))
        {
            throw new NotFoundException(NotFoundException.NotFoundReason.UserNotFound);

        }

        var cartId = _cartIdCounter++;
        var cart = new Cart
        {
            CartId = cartId,
            UserId = request.UserId,
            Items = new List<CartItem>()
        };

        Carts[cartId] = cart;

        return new InitializeCartResponse { CartId = cartId };
    }

    public void AddItemToCart(uint cartId, AddItemToCartRequest request)
    {
        if (!Carts.TryGetValue(cartId, out var cart))
        {
            throw new NotFoundException(NotFoundException.NotFoundReason.CartNotFound);
        }

        if (!_productService.Exists(request.ProductId))
        {
            throw new NotFoundException(NotFoundException.NotFoundReason.ProductNotFound);
        }

        if (cart.Items.Any(i => i.ProductId == request.ProductId))
        {
            throw new ConflictException($"Product {request.ProductId} already exists in cart");
        }

        cart.Items.Add(new CartItem
        {
            ProductId = request.ProductId,
            Quantity = request.Quantity
        });
    }

    public Cart GetCart(uint cartId)
    {
        if (!Carts.TryGetValue(cartId, out var cart))
        {
            throw new NotFoundException(NotFoundException.NotFoundReason.CartNotFound);
        }
        return cart;
    }
}

File: consumer/src/services/cart-service.ts
import axios from 'axios';
import { AddItemToCartRequest } from '../entities/add-item-to-cart-request';
import { AddItemToCartNotFoundResponse } from '../entities/add-item-to-cart-notfound-response';
import { Cart } from '../entities/cart';
import { InitializeCartRequest } from '../entities/initialize-cart-request';
import { InitializeCartResponse } from '../entities/initialize-cart-response';

export class CartService {
    readonly baseUrl: string;

    constructor(baseUrl: string = 'http://localhost:9000') {
        this.baseUrl = baseUrl;
    }

    async initializeCart(request: InitializeCartRequest): Promise<InitializeCartResponse> {
        try {
            const { data } = await axios.post<InitializeCartResponse>(`${this.baseUrl}/carts`, request);
            return data;
        } catch (error) {
            if (axios.isAxiosError(error)) {
                throw new Error(`Failed to initialize cart: ${error.response?.status}`);
            }
            throw error;
        }
    }

    async addItemToCart(cartId: number, request: AddItemToCartRequest): Promise<void | AddItemToCartNotFoundResponse> {
        try {
            await axios.post(`${this.baseUrl}/carts/${cartId}/items`, request);
        } catch (error) {
            if (axios.isAxiosError(error)) {
                if (error.response?.status === 404) {
                    return error.response.data as AddItemToCartNotFoundResponse;
                }
                if (error.response?.status === 409) {
                    throw new Error('Item already in cart');
                }
                throw new Error(`Failed to add item to cart: ${error.response?.status}`);
            }
            throw error;
        }
    }

    async getCart(cartId: number): Promise<Cart> {
        try {
            const { data } = await axios.get<Cart>(`${this.baseUrl}/carts/${cartId}`);
            return data;
        } catch (error) {
            if (axios.isAxiosError(error)) {
                throw new Error('Cart not found');
            }
            throw error;
        }
    }
}